components/raftstore-v2/src/router/response_channel.rs

```diff
@@ -0,0 +1,448 @@
+// Copyright 2022 TiKV Project Authors. Licensed under Apache-2.0.
+
+//! Variants of channels for `Msg`.
+//!  - `Read`: a channel for read only requests including `StatusRequest`,
+//!    `GetRequest` and `SnapRequest`
+//!  - `Write`: a channel for write only requests including `AdminRequest`
+//!    `PutRequest`, `DeleteRequest` and `DeleteRangeRequest`.
+//!
+//! Prefer channel over callback because:
+//! 1. channel can be reused, hence reduce allocations (not yet implemented).
+//! 2. channel may not need dynamic dispatch.
+//! 3. caller can use async fashion.
+//! 4. there will be no callback leak.
+
+use std::{
+    cell::UnsafeCell,
+    fmt,
+    future::Future,
+    mem::{self, ManuallyDrop},
+    pin::Pin,
+    sync::{
+        atomic::{AtomicU64, Ordering},
+        Arc,
+    },
+    task::{Context, Poll},
+};
+
+use engine_traits::Snapshot;
+use futures::task::AtomicWaker;
+use kvproto::{kvrpcpb::ExtraOp as TxnExtraOp, raft_cmdpb::RaftCmdResponse};
+use raftstore::store::{
+    local_metrics::TimeTracker, msg::ErrorCallback, ReadCallback, RegionSnapshot, WriteCallback,
+};
+use smallvec::SmallVec;
+use tikv_util::memory::HeapSize;
+
+/// A struct allows to watch and notify specific events.
+///
+/// There are two different events: state and payload. Obviously, state events
+/// have no payload. At most 30 states can be defined. There can be only one
+/// type of payload.
+struct EventCore<Res> {
+    /// Every event will have two bits.
+    /// - 0b00 means the event is not fired and not watched.
+    /// - 0b01 means the event is fired and not watched.
```

Comment:
define these 0bxx as more readable const?

Reply:
I think it will be better to define the event enum so that we can index all kinds of events easily in one place and it's safer. 
Also the logic of bit operation can be wrapped into that enum. Instead of using subscribed_bit_of and set_bit_of in notify_event, we can use more user friendly name such as fire(), is_subscribed(). Even though the notify_event is not called directly. It's still useful to make code more readable.

Reply:
How do you apply atomic operation on enum?